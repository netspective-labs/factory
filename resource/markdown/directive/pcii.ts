import * as c from "../../content/mod.ts";
import * as mdDS from "../mod.ts";
import * as extn from "../../../module/mod.ts";

export interface ProxiedContentIIDirectiveAttrs {
  readonly module?: string;
}

export class ProxiedContentInfuseInterpolateDirective
  implements
    c.DirectiveExpectation<
      mdDS.MarkdownContentInlineDirective<ProxiedContentIIDirectiveAttrs>,
      string | undefined
    > {
  static readonly IDENTITY = "pcii";
  readonly identity = ProxiedContentInfuseInterpolateDirective.IDENTITY;

  constructor(readonly extnManager: extn.ExtensionsManager) {
  }

  text(
    symbol: string,
    extn: extn.ExtensionModule,
    directive: mdDS.MarkdownContentInlineDirective<
      ProxiedContentIIDirectiveAttrs
    >,
  ): undefined | string {
    let found = false;
    let symbolValue: unknown = undefined;
    extn.exports((key, value) => {
      if (key == symbol) {
        if (!found) {
          found = true;
          symbolValue = value;
          return true;
        }
      }
      return false;
    });
    if (!found) return undefined;

    switch (typeof symbolValue) {
      case "function":
        return `${symbolValue(directive)}`;

      default:
        return `${symbolValue}`;
    }
  }

  encountered(
    d: mdDS.MarkdownContentInlineDirective<ProxiedContentIIDirectiveAttrs>,
  ): string | undefined {
    const symbol = d.content;
    if (!symbol || symbol.length == 0) {
      return `PCII symbol not declared`;
    }

    if (symbol) {
      const searched = [];
      if (d.attributes?.module) {
        const module = d.attributes?.module;
        for (const extn of this.extnManager.extensions) {
          if (extn.provenance.toString().endsWith(module)) {
            const text = this.text(symbol, extn, d);
            if (text) return text;
            return `PCII symbol '${symbol}' was not found in module '${module}' (make sure it's defined and exported)`;
          }
          searched.push(extn.provenance);
        }
        // we cannot import a module since Markdown directives are synchronous-only (no async allowed)
        return `PCII module '${module}' was not found (be sure to import it into extensions manager), searched: ${
          searched.join(", ")
        }`;
      }

      for (const extn of this.extnManager.extensions) {
        const text = this.text(symbol, extn, d);
        if (text) return text;
        searched.push(extn.provenance);
      }
      return `PCII symbol '${symbol}' was not found in any module (make sure it's defined and exported in an imported module)`;
    }

    // TODO: support globals in window.*
    return `unknown PCII origin`;
  }
}
