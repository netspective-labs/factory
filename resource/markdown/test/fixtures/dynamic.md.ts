import * as r from "./synthetic.pcii.ts";

// in *.md.ts all exported symbols are considered "typed" frontmatter
export const title = "Dynamic Markdown";

// the default text is considered the markdown content and may be completely dynamic
// deno-fmt-ignore
export default `

${r.var1}

# Dynamic Heading

- dynamic bullet #1
- dynamic bullet #2
`;
