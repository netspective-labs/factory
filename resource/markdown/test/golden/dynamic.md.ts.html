<div class="md"><p>Sample</p>
<h1 id="dynamic-heading" tabindex="-1">Dynamic Heading</h1>
<ul>
<li>dynamic bullet <a href="/tags/1" class="tag">#1</a></li>
<li>dynamic bullet <a href="/tags/2" class="tag">#2</a></li>
</ul>
</div>